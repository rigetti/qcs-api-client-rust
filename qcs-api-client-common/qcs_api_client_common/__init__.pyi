from . import configuration

__all__ = ["configuration"]
