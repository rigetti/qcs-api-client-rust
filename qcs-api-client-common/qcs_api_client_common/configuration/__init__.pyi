from typing import Callable, final

DEFAULT_API_URL: str
DEFAULT_GRPC_API_URL: str
DEFAULT_QUILC_URL: str
DEFAULT_QVM_URL: str
DEFAULT_PROFILE_NAME: str
PROFILE_NAME_VAR: str
QUILC_URL_VAR: str
QVM_URL_VAR: str
API_URL_VAR: str
GRPC_API_URL_VAR: str
SETTINGS_PATH_VAR: str
SECRETS_PATH_VAR: str
DEFAULT_SETTINGS_PATH: str
DEFAULT_SECRETS_PATH: str

@final
class ClientConfiguration:
    """A configuration suitable for use as a QCS API Client.

    This configuration can be constructed in a few ways.

    The most common way is to use `ClientConfiguration#load_default`. This will load the
    configuration associated with your default QCS profile. When loading a profile, any values set by environment variables will override the values in your configuration files.

    You can also build a configuration from scratch using `ClientConfigurationBuilder`. Using a
    builder bypasses configuration files and environment overrides.
    """
    @staticmethod
    def load_default() -> ClientConfiguration:
        """Load a `ClientConfiguration` using your default QCS profile."""
        ...

    @staticmethod
    def load_profile(profile_name: str) -> ClientConfigurationBuilder:
        """Load a `ClientConfiguration` using the given QCS profile."""

    @staticmethod
    def builder() -> ClientConfigurationBuilder:
        """Create a new `ClientConfigurationBuilder`."""

    @property
    def api_url(self) -> str:
        """The URL of the QCS REST API."""

    @property
    def grpc_api_url(self) -> str:
        """The URL of the QCS gRPC API."""

    @property
    def quilc_url(self) -> str:
        """The URL of the QCS quilc API."""

    @property
    def qvm_url(self) -> str:
        """The URL of the QCS QVM API."""

    def get_oauth_session(self) -> OAuthSession:
        """Get the credentials used to authenticate with the QCS API."""

    async def get_oauth_session_async(self) -> OAuthSession:
        """Get the credentials used to authenticate with the QCS API."""

    def get_bearer_access_token(self) -> str:
        """Gets the `Bearer` access token, refreshing it if is expired."""

    async def get_bearer_access_token_async(self) -> str:
        """Gets the `Bearer` access token, refreshing it if is expired."""

@final
class ClientConfigurationBuilder:
    def __new__(cls) -> ClientConfigurationBuilder: ...
    def build(self) -> ClientConfiguration:
        """Build a `ClientConfiguration` using the values provided to this builder."""
        ...

    @property
    def api_url(self):
        raise AttributeError("api_url is not readable")

    @api_url.setter
    def api_url(self, api_url: str):
        """Set the URL to use for the QCS REST API."""

    @property
    def grpc_api_url(self):
        raise AttributeError("grpc_api_url is not readable")

    @grpc_api_url.setter
    def grpc_api_url(self, grpc_api_url: str):
        """Set the URL to use for the QCS gRPC API."""

    @property
    def quilc_url(self):
        raise AttributeError("quilc_url is not readable")

    @quilc_url.setter
    def quilc_url(self, quilc_url: str):
        """Set the URL to use for the quilc server."""

    @property
    def qvm_url(self):
        raise AttributeError("qvm_url is not readable")

    @qvm_url.setter
    def qvm_url(self, qvm_url: str):
        """Set the URL to use for the QVM server."""

    @property
    def oauth_session(self):
        raise AttributeError("tokens is not readable")

    @oauth_session.setter
    def oauth_session(self, tokens: OAuthSession | None):
        """Set the QCS API access and refresh `Credentials` to use."""

@final
class AuthServer:
    def __new__(cls, client_id: str, issuer: str) -> AuthServer: ...
    @staticmethod
    def default() -> AuthServer:
        """Get the default Okta auth server."""

    @property
    def client_id(self) -> str:
        """The client's Okta ID."""

    @property
    def issuer(self) -> str:
        """The Okta issuer URL."""

@final
class RefreshToken:
    def __new__(cls, refresh_token: str) -> RefreshToken: ...
    @property
    def refresh_token(self) -> str:
        """The refresh token."""
    @refresh_token.setter
    def refresh_token(self, refresh_token: str):
        """Set the refresh token."""

@final
class ClientCredentials:
    def __new__(cls, client_id: str, client_secret: str) -> ClientCredentials: ...
    @property
    def client_id(self) -> str:
        """The client ID."""
    @property
    def client_secret(self) -> str:
        """The client secret."""

@final
class ExternallyManaged:
    def __new__(cls, refresh_function: Callable[[AuthServer], str]) -> ExternallyManaged:
        """Manages access tokens by utilizing a user-provided refresh function.

        The refresh function should return a valid access token, or raise an exception if it cannot.

        .. testcode::
            from qcs_apiclient_common.configuration import AuthServer, ExternallyManaged, OAuthSession

            def refresh_function(auth_server: AuthServer) -> str:
                return "my_access_token"

            externally_managed = ExternallyManaged(refresh_function)
            session = OAuthSession(externally_managed, AuthServer.default())
        """

@final
class OAuthSession:
    def __new__(
        cls,
        grant_payload: RefreshToken | ClientCredentials | ExternallyManaged,
        auth_server: AuthServer,
        access_token: str | None = None,
    ) -> OAuthSession: ...
    @property
    def access_token(self) -> str:
        """Get the current access token.

        This is an unvalidated copy of the access token. Meaning it can become stale, or may already be stale. See the `validate` `request_access_token` and methods.
        """

    @property
    def auth_server(self) -> AuthServer:
        """The auth server."""

    @property
    def payload(self) -> RefreshToken | ClientCredentials:
        """Get the payload used to request an access token."""

    def request_access_token(self) -> str:
        """Request a new access token."""

    async def request_access_token_async(self) -> str:
        """Request a new access token."""

    def validate(self) -> str:
        """Validate the current access token, returning it if it is valid.

        If the token is invalid, a `ValueError` will be raised with a description of why the token failed validation.
        """
