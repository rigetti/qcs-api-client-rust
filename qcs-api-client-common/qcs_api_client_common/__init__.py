""".. include:: ../README-py.md""" # noqa: D415
from .qcs_api_client_common import configuration

__all__ = ["configuration"]
