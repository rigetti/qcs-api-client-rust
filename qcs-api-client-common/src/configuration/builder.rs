use std::sync::Arc;

use thiserror::Error;
use tokio::sync::Mutex;

use crate::ClientConfiguration;

use super::{
    settings::AuthServer, Tokens, DEFAULT_API_URL, DEFAULT_GRPC_API_URL, DEFAULT_QUILC_URL,
    DEFAULT_QVM_URL, GRPC_API_URL_VAR, QUILC_URL_VAR, QVM_URL_VAR,
};

#[cfg(feature = "tracing-config")]
use crate::tracing_configuration::TracingConfiguration;

/// Errors that may occur when building a [`ClientConfiguration`].
#[derive(Debug, Error)]
#[non_exhaustive]
pub enum BuildError {}

/// Builder for [`ClientConfiguration`] to set/override items.
#[allow(clippy::module_name_repetitions)]
#[derive(Debug, Default, Clone)]
pub struct ClientConfigurationBuilder {
    tokens: Option<Arc<Mutex<Tokens>>>,
    api_url: Option<String>,
    auth_server: Option<AuthServer>,
    grpc_api_url: Option<String>,
    quilc_url: Option<String>,
    qvm_url: Option<String>,
    #[cfg(feature = "tracing-config")]
    tracing_configuration: Option<TracingConfiguration>,
}

impl From<ClientConfiguration> for ClientConfigurationBuilder {
    fn from(config: ClientConfiguration) -> Self {
        Self {
            tokens: Some(config.tokens),
            api_url: Some(config.api_url),
            auth_server: Some(config.auth_server),
            grpc_api_url: Some(config.grpc_api_url),
            quilc_url: Some(config.quilc_url),
            qvm_url: Some(config.qvm_url),
            #[cfg(feature = "tracing-config")]
            tracing_configuration: config.tracing_configuration,
        }
    }
}

impl ClientConfigurationBuilder {
    #![allow(clippy::missing_const_for_fn)]

    /// Set the [`Tokens`] used for authentication.
    #[must_use]
    pub fn set_tokens(self, tokens: Tokens) -> Self {
        self.set_tokens_arc(Arc::new(Mutex::new(tokens)))
    }

    /// Reuse the given [`Tokens`] for authentication.
    #[must_use]
    pub fn set_tokens_arc(mut self, tokens: Arc<Mutex<Tokens>>) -> Self {
        self.tokens = Some(tokens);
        self
    }

    /// Set the OpenAPI URL to use.
    #[must_use]
    pub fn set_api_url(mut self, url: String) -> Self {
        self.api_url = Some(url);
        self
    }

    /// Set the authentication server URL to use.
    #[must_use]
    pub fn set_auth_server(mut self, auth_server: AuthServer) -> Self {
        self.auth_server = Some(auth_server);
        self
    }

    /// Set the gRPC API URL to use.
    #[must_use]
    pub fn set_grpc_api_url(mut self, url: String) -> Self {
        self.grpc_api_url = Some(url);
        self
    }

    /// Set the Quilc HTTP server to use.
    #[must_use]
    pub fn set_quilc_url(mut self, url: String) -> Self {
        self.quilc_url = Some(url);
        self
    }

    /// Set the QVM HTTP server to use.
    #[must_use]
    pub fn set_qvm_url(mut self, url: String) -> Self {
        self.qvm_url = Some(url);
        self
    }

    /// Set the [`TracingConfiguration`]. If set to `None`, network API calls will not be traced.
    /// Otherwise, the given [`TracingConfiguration`] will be used to configure tracing.
    #[cfg(feature = "tracing-config")]
    #[must_use]
    pub fn set_tracing_configuration(
        mut self,
        tracing_configuration: Option<TracingConfiguration>,
    ) -> Self {
        self.tracing_configuration = tracing_configuration;
        self
    }

    /// Build the [`ClientConfiguration`].
    ///
    /// # Errors
    ///
    /// See [`BuildError`].
    pub fn build(self) -> Result<ClientConfiguration, BuildError> {
        Ok(ClientConfiguration {
            tokens: self.tokens.unwrap_or_default(),
            api_url: self
                .api_url
                .or_else(|| std::env::var(API_URL_VAR).ok())
                .unwrap_or_else(|| DEFAULT_API_URL.to_string()),
            auth_server: self.auth_server.unwrap_or_default(),
            grpc_api_url: self
                .grpc_api_url
                .or_else(|| std::env::var(GRPC_API_URL_VAR).ok())
                .unwrap_or_else(|| DEFAULT_GRPC_API_URL.to_string()),
            quilc_url: self
                .quilc_url
                .or_else(|| std::env::var(QUILC_URL_VAR).ok())
                .unwrap_or_else(|| DEFAULT_QUILC_URL.to_string()),
            qvm_url: self
                .qvm_url
                .or_else(|| std::env::var(QVM_URL_VAR).ok())
                .unwrap_or_else(|| DEFAULT_QVM_URL.to_string()),
            #[cfg(feature = "tracing-config")]
            tracing_configuration: self.tracing_configuration,
        })
    }
}

#[cfg(test)]
#[allow(clippy::cognitive_complexity)]
mod tests {
    use super::*;
    use rstest::rstest;
    use serial_test::serial;

    fn get_tokens() -> Tokens {
        Tokens {
            bearer_access_token: String::from("custom access"),
            refresh_token: String::from("custom refresh"),
        }
    }

    fn get_tokens_arc() -> Arc<Mutex<Tokens>> {
        Arc::new(Mutex::new(Tokens {
            bearer_access_token: String::from("custom access"),
            refresh_token: String::from("custom refresh"),
        }))
    }

    fn get_auth_server() -> AuthServer {
        AuthServer::default()
            .set_client_id(String::from("fake client id"))
            .set_issuer(String::from("fake issuer"))
    }

    macro_rules! set_from_option {
        ($builder: ident, $set_fn: ident, $op: ident) => {
            if let Some(val) = $op.clone() {
                $builder = $builder.$set_fn(val);
            }
        };
    }

    macro_rules! assert_set_from_option {
        ($field: expr, $op: ident, $default: expr) => {
            if let Some(val) = $op {
                assert_eq!($field, val, "expected value to be set in builder");
            } else {
                assert_eq!($field, $default, "expected default value to be used");
            }
        };
    }

    #[rstest]
    #[serial]
    fn test_builder_sets_options_or_default(
        #[values(None, Some(get_tokens()))] tokens: Option<Tokens>,
        #[values(None, Some(get_tokens_arc()))] tokens_arc: Option<Arc<Mutex<Tokens>>>,
        #[values(None, Some(String::from("custom api url")))] api_url: Option<String>,
        #[values(None, Some(get_auth_server()))] auth_server: Option<AuthServer>,
        #[values(None, Some(String::from("custom grpc url")))] grpc_api_url: Option<String>,
        #[values(None, Some(String::from("custom quilc url")))] quilc_url: Option<String>,
        #[values(None, Some(String::from("custom qvm url")))] qvm_url: Option<String>,
    ) {
        let mut builder = ClientConfigurationBuilder::default();

        std::env::remove_var(QUILC_URL_VAR);
        std::env::remove_var(QVM_URL_VAR);
        std::env::remove_var(GRPC_API_URL_VAR);

        set_from_option!(builder, set_tokens, tokens);
        set_from_option!(builder, set_tokens_arc, tokens_arc);
        set_from_option!(builder, set_api_url, api_url);
        set_from_option!(builder, set_auth_server, auth_server);
        set_from_option!(builder, set_grpc_api_url, grpc_api_url);
        set_from_option!(builder, set_quilc_url, quilc_url);
        set_from_option!(builder, set_qvm_url, qvm_url);

        let config = builder.build().unwrap();

        {
            // Handle tokens specially
            let expected_tokens = tokens_arc.map(|arc| arc.blocking_lock().clone()).or(tokens);
            let actual_tokens = Arc::try_unwrap(config.tokens).unwrap().into_inner();
            assert_set_from_option!(actual_tokens, expected_tokens, Tokens::default());
        }

        assert_set_from_option!(config.api_url, api_url, DEFAULT_API_URL);
        assert_set_from_option!(config.auth_server, auth_server, AuthServer::default());
        assert_set_from_option!(config.grpc_api_url, grpc_api_url, DEFAULT_GRPC_API_URL);
        assert_set_from_option!(config.quilc_url, quilc_url, DEFAULT_QUILC_URL);
        assert_set_from_option!(config.qvm_url, qvm_url, DEFAULT_QVM_URL);
    }
}
