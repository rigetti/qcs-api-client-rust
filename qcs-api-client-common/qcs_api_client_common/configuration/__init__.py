"""This module facilitates loading a QCS configuration to connect to real QPUs or the QVM.

The primary class provided is :class:`ClientConfiguration`, which allows you to load and access your configuration.
You can load the configuration from disk, typically from the ``~/.qcs`` directory (recommended for most users), or programmatically using
:class:`ClientConfigurationBuilder`.

### Loading Configuration from Disk

Most users will only need to use :meth:`ClientConfiguration.load_default`. This method instantiates a new
:class:`ClientConfiguration` with settings loaded from files in the ``~/.qcs`` directory, including:

* ``settings.toml``: General settings, such as connection URLs.
* ``secrets.toml``: Authentication tokens.

Both files should contain profiles. The ``default_profile_name`` setting determines which profile is loaded by default.
To load a non-default profile, use :meth:`ClientConfiguration.load_profile`.

For details on obtaining these files, refer to the `QCS Credentials Guide <https://docs.rigetti.com/qcs/guides/qcs-credentials>`_.

### Using the ClientConfigurationBuilder

:class:`ClientConfigurationBuilder` allows for programmatically building a :class:`ClientConfiguration`. This is useful
for loading configurations from alternative sources or overriding specific settings. Advanced functionality, such as
using a different OAuth grant type, is only available through the builder.

Example usage:

.. code-block:: python

    from qcs_api_client_common.configuration import ClientConfiguration, OAuthSession, ClientCredentials

    # Initialize a builder
    builder = ClientConfiguration.builder()

    # Set custom settings, e.g., a custom OAuth session
    oauth = OAuthSession(ClientCredentials("client_id", "client_secret"), AuthServer.default())
    builder.oauth_session(oauth)

    # Build and retrieve the configuration
    configuration = builder.build()

### Environment Variables

You can override configuration values using environment variables, which take precedence over values loaded from files
but are overridden by settings explicitly set in a builder.

* ``QCS_SETTINGS_FILE_PATH``: Path to the ``settings.toml`` file.
* ``QCS_SECRETS_FILE_PATH``: Path to the ``secrets.toml`` file.
* ``QCS_PROFILE_NAME``: Profile to load from ``settings.toml``.
* ``QCS_SETTINGS_APPLICATIONS_QUILC_URL``: URL for the quilc server.
* ``QCS_SETTINGS_APPLICATIONS_QVM_URL``: URL for the QVM server.
* ``QCS_SETTINGS_APPLICATIONS_API_URL``: URL for the QCS REST API server.
* ``QCS_SETTINGS_APPLICATIONS_GRPC_URL``: URL for the QCS gRPC API.

See the `QCS Client Configuration Guide <https://docs.rigetti.com/qcs/references/qcs-client-configuration>`_ for more details.
"""
