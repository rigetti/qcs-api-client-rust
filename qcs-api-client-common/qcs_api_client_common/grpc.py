"""Middleware for gRPC clients."""
from abc import ABCMeta
from typing import Any, Callable, List, Optional, Tuple

import grpc  # type: ignore
from grpc import ClientCallDetails  # type: ignore
from grpc.aio import Call, UnaryUnaryClientInterceptor  # type: ignore

from qcs_api_client_common.configuration import ClientConfiguration


class RefreshInterceptor(UnaryUnaryClientInterceptor, metaclass=ABCMeta):
    """A `RefreshInterceptor` will add your QCS authorization token to all your QCS requests.

    The interceptor will automatically refresh your token as needed.
    """
    def __init__(self, client_configuration: Optional[ClientConfiguration]=None):
        """Initialize the interceptor using a `ClientConfiguration`.

        If `client_configuration` is unset, `ClientConfiguratio.load_default()` will be used.
        """
        if client_configuration is None:
            client_configuration = ClientConfiguration.load_default()
        self.configuration = client_configuration

    async def _get_access_token(self) -> str:
        return await self.configuration.get_bearer_access_token_async()

    async def intercept_unary_unary(
        self,
        continuation: Callable[[ClientCallDetails, Any], Call],
        client_call_details: ClientCallDetails,
        request: Any
    ) -> Any:
        """Adds the QCS authorization token to the request metadata, refreshing the token if needed."""
        # Modify the headers to add the access token
        authorized_metadata: List[Tuple[str, str]] = []
        if client_call_details.metadata is not None:
            authorized_metadata = list(client_call_details.metadata)

        authorized_metadata.append(('authorization', f'Bearer {await self._get_access_token()}'))

        new_client_call_details = grpc.aio.ClientCallDetails(
            client_call_details.method,
            client_call_details.timeout,
            authorized_metadata,
            client_call_details.credentials,
            client_call_details.wait_for_ready,
        )

        return await continuation(new_client_call_details, request)
