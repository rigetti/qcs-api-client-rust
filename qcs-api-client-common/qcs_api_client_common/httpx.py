"""QCS Middleware for requests made with `httpx`."""

from typing import Optional

import httpx
from httpx import Request

from qcs_api_client_common.configuration import ClientConfiguration


class QCSAuthorization(httpx.Auth):
    """An `httpx.Auth` that will add QCS authorization to your requests and refresh the token as needed.

    Synchronous usage:

    ```python
    import httpx

    auth = QCSAuthorization()
    client = httpx.Client(auth=auth)

    response = client.get('https://rigetti.com/protected-resource')
    ```

    Asynchronous usage:

    ```python
    import httpx
    import asyncio

    async def main():
        auth = QCSAuthorization()
        async with httpx.AsyncClient(auth=auth) as client:
            response = await client.get('https://rigetti.com/protected-resource')
            print(response.status_code)
            print(response.json())

    asyncio.run(main())
    ```
    """
    def __init__(self, client_configuration: Optional[ClientConfiguration] = None):
        """Initialize the authorization with an optional client configuration.

        If `client_configuration` is set to None, `ClientConfiguration.load_default()` will be used.
        """
        self.configuration = client_configuration or ClientConfiguration.load_default()

    # Note: No explicit type hints here. It confuses stubtest, and they are inherited from the base class anyways.
    async def async_auth_flow(self, request: Request):
        """Add the QCS authorization token to the request, refreshing the token as needed."""
        token = await self.configuration.get_bearer_access_token_async()
        request.headers['Authorization'] = f'Bearer {token}'
        yield request


    def sync_auth_flow(self, request: Request):
        """Add the QCS authorization token to the request, refreshing the token as needed."""
        token = self.configuration.get_bearer_access_token()
        request.headers['Authorization'] = f'Bearer {token}'
        yield request
