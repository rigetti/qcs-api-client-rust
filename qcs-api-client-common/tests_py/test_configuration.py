from syrupy.assertion import SnapshotAssertion

from qcs_api_client_common.configuration import AuthServer, ClientConfiguration, OAuthSession, RefreshToken


class TestClientConfiguration:
    def test_default(self, snapshot: SnapshotAssertion):
        cc = ClientConfiguration.load_default()
        assert cc == snapshot

    def test_load_profile(self, snapshot: SnapshotAssertion):
        cc = ClientConfiguration.load_profile("test")
        assert cc == snapshot


class TestClientConfigurationBuilder:
    def test_build(self, snapshot: SnapshotAssertion):
        builder = ClientConfiguration.builder()
        builder.api_url = "builder_api_url"
        builder.grpc_api_url = "builder_grpc_api_url"
        builder.quilc_url = "builder_quilc_url"
        builder.qvm_url = "builder_qvm_url"
        auth_server = AuthServer("builder_client_id", "builder_issuer")
        builder.oauth_session = OAuthSession(RefreshToken("builder_refresh_token"), auth_server, "builder_access_token")
        assert builder.build() == snapshot


class TestAuthServer:
    def test_properties(self):
        auth_server = AuthServer("client_id", "issuer")
        assert auth_server.client_id == "client_id"
        assert auth_server.issuer == "issuer"

    def test_eq(self):
        auth_server = AuthServer("client_id", "issuer")
        assert auth_server == auth_server
        assert auth_server == AuthServer("client_id", "issuer")
        assert auth_server != AuthServer("different_client_id", "issuer")
        assert auth_server != AuthServer("client_id", "different_issuer")


class TestCredentials:
    def test_properties(self):
        payload = RefreshToken("refresh")
        auth_server = AuthServer("some_client_id", "some_issuer")
        credentials = OAuthSession(payload, auth_server, "access")
        assert credentials.access_token == "access"
        assert credentials.auth_server == auth_server
        assert credentials.payload == payload

    def test_eq(self):
        payload = RefreshToken("refresh")
        auth_server = AuthServer("some_client_id", "some_issuer")
        credentials = OAuthSession(payload, auth_server, "access")
        assert credentials == credentials
        assert credentials == OAuthSession(payload, auth_server, "access")
        assert credentials != OAuthSession(payload, auth_server, "different_access")
        assert credentials != OAuthSession(RefreshToken("different_refresh"), auth_server, "access")
