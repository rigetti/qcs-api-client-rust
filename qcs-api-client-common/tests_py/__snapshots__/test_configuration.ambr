# serializer version: 1
# name: TestClientConfiguration.test_default
  ClientConfiguration { profile: "production", api_url: "https://api.qcs.rigetti.com", grpc_api_url: "https://grpc.qcs.rigetti.com", quilc_url: "tcp://127.0.0.1:5555", qvm_url: "http://127.0.0.1:5000", oauth_session: Some(TokenDispatcher { lock: RwLock { data: OAuthSession { payload: RefreshToken(RefreshToken { refresh_token: "refresh_production" }), access_token: None, auth_server: AuthServer { client_id: "client_id_production", issuer: "https://auth.qcs.rigetti.com/oauth2/production" } } }, refreshing: Mutex { data: false }, notify_refreshed: Notify { state: 0, waiters: Mutex(Mutex { data: LinkedList { head: None, tail: None }, poisoned: false, .. }) } }) }
# ---
# name: TestClientConfiguration.test_load_profile
  ClientConfiguration { profile: "test", api_url: "https://test.qcs.rigetti.com", grpc_api_url: "https://grpc.test.qcs.rigetti.com", quilc_url: "tcp://127.0.0.1:5555", qvm_url: "http://127.0.0.1:5000", oauth_session: Some(TokenDispatcher { lock: RwLock { data: OAuthSession { payload: RefreshToken(RefreshToken { refresh_token: "refresh_test" }), access_token: None, auth_server: AuthServer { client_id: "client_id_test", issuer: "https://auth.test.qcs.rigetti.com/oauth2/test" } } }, refreshing: Mutex { data: false }, notify_refreshed: Notify { state: 0, waiters: Mutex(Mutex { data: LinkedList { head: None, tail: None }, poisoned: false, .. }) } }) }
# ---
# name: TestClientConfigurationBuilder.test_build
  ClientConfiguration { profile: "default", api_url: "builder_api_url", grpc_api_url: "builder_grpc_api_url", quilc_url: "builder_quilc_url", qvm_url: "builder_qvm_url", oauth_session: Some(TokenDispatcher { lock: RwLock { data: OAuthSession { payload: RefreshToken(RefreshToken { refresh_token: "builder_refresh_token" }), access_token: Some("builder_access_token"), auth_server: AuthServer { client_id: "builder_client_id", issuer: "builder_issuer" } } }, refreshing: Mutex { data: false }, notify_refreshed: Notify { state: 0, waiters: Mutex(Mutex { data: LinkedList { head: None, tail: None }, poisoned: false, .. }) } }) }
# ---
